main/say.c

```diff
@@ -353,26 +354,45 @@ static int say_digit_str_full(struct ast_channel *chan, const char *str, const c
 static struct ast_str* ast_get_money_en_dollars_str(const char *str, const char *lang)
 {
 	const char *fnr;
-
-	double dollars = 0;
-	int amt, cents;
+	const char *period;
+	int amt, dollars, cents = 0;
 	struct ast_str *fnrecurse = NULL;
+	struct ast_str *filenames;
 
-	struct ast_str *filenames = ast_str_create(20);
+	if (ast_strlen_zero(str)) {
+		return NULL;
+	}
+
+	filenames = ast_str_create(20);
 	if (!filenames) {
 		return NULL;
 	}
 	ast_str_reset(filenames);
 
-	if (sscanf(str, "%30lf", &dollars) != 1) {
-		amt = 0;
-	} else { /* convert everything to cents */
-		amt = dollars * 100;
+	/* Don't use sscanf because floating point rounding
+	 * could distort the cents units. Just parse as string. */
+	dollars = atoi(str); /* atoi will stop at the period, if there is one */
+	period = strchr(str, '.');
+	if (period) {
```

Comment:
The decimal point would need to be just after the initial integer sequence, otherwise this would accept something like `2non-digit-stuff.05` as `2.05` which seems wrong (and regardless of FP accuracy is not consistent with the current behavior).

Using `strtol(...)` instead of `atoi(...)` would give you a pointer to the first character after any initial digits and you could just check that directly against `'.'`.

Reply:
And now that I think about it I think this would accept `blahblah.05` as `0.05` which also seems incorrect.

Reply:
I think I was overthinking this big time, I redid the patch with simpler logic that is more correct